src/Compilers/CSharp/Portable/Parser/LanguageParser_InterpolatedString.cs

```diff
@@ -51,142 +34,174 @@ private ExpressionSyntax ParseInterpolatedStringToken()
             // 
             // The substitution will end up being invisible to external APIs and clients such as the IDE, as
             // they have no way to ask for the stream of tokens before parsing.
-            //
 
+            Debug.Assert(this.CurrentToken.Kind == SyntaxKind.InterpolatedStringToken);
             var originalToken = this.EatToken();
+
             var originalText = originalToken.ValueText; // this is actually the source text
             Debug.Assert(originalText[0] == '$' || originalText[0] == '@');
 
-            var isAltInterpolatedVerbatim = originalText.Length > 2 && originalText[0] == '@'; // @$
-            var isVerbatim = isAltInterpolatedVerbatim || (originalText.Length > 2 && originalText[1] == '@');
+            var isVerbatim = (originalText[0] == '$' && originalText[1] == '@') ||
+                             (originalText[0] == '@' && originalText[1] == '$');
 
-            Debug.Assert(originalToken.Kind == SyntaxKind.InterpolatedStringToken);
+            // compute the positions of the interpolations in the original string literal, if there was an error or not,
+            // and where the close quote can be found.
             var interpolations = ArrayBuilder<Lexer.Interpolation>.GetInstance();
-            SyntaxDiagnosticInfo error;
-            bool closeQuoteMissing;
-            using (var tempLexer = new Lexer(Text.SourceText.From(originalText), this.Options, allowPreprocessorDirectives: false))
+
+            rescanInterpolation(out var openQuoteRange, out var error, out var closeQuoteRange);
+
+            var result = SyntaxFactory.InterpolatedStringExpression(
+                getOpenQuote(openQuoteRange), getContent(interpolations), getCloseQuote(closeQuoteRange));
+
+            interpolations.Free();
+            if (error != null)
+            {
+                result = result.WithDiagnosticsGreen(new[] { error });
+            }
+
+            Debug.Assert(originalToken.ToFullString() == result.ToFullString()); // yield from text equals yield from node
+            return result;
+
+            void rescanInterpolation(out Range openQuoteRange, out SyntaxDiagnosticInfo error, out Range closeQuoteRange)
             {
-                // compute the positions of the interpolations in the original string literal, and also compute/preserve
-                // lexical errors
+                using var tempLexer = new Lexer(SourceText.From(originalText), this.Options, allowPreprocessorDirectives: false);
                 var info = default(Lexer.TokenInfo);
-                tempLexer.ScanInterpolatedStringLiteralTop(interpolations, isVerbatim, ref info, out error, out closeQuoteMissing);
+                tempLexer.ScanInterpolatedStringLiteralTop(ref info, out error, out openQuoteRange, interpolations, out closeQuoteRange);
```

Comment:
>interpolations [](http://example.com/codeflow?start=100&length=14)

Is this the only thing that this local function needs from the surrounding context? If so, consider making the function static and passing this in. It's far enough away from the usage that it's not immediately obvious that interpolations is used in here when reading the main method body, and since the builder is immediately freed there's some refactoring risk there.

Reply:
yeah, i wasn't sure the best structure here.  i've changed it uip a bit now to hopefully make it a bit clearer.  now the data that is in/out should be more obvious/relevant (i think).